### PR TITLE
Feature/password slot protection #312

### PR DESCRIPTION
**Pull request by va-dym-ko-val**
image
image
image

### CODE REVIEW

File: frontend/src/app/modules/external-booking/external-booking-choose-time-page/external-booking-choose-time-page.component.ts [hunk 1]
```diff
@@ -19,7 +21,7 @@ import { TZone } from 'moment-timezone-picker';
     templateUrl: './external-booking-choose-time-page.component.html',
     styleUrls: ['./external-booking-choose-time-page.component.sass'],
 })
-export class ExternalBookingTimeComponent extends BaseComponent implements OnInit {
+export class ExternalBookingTimeComponent extends BaseComponent implements OnInit, OnDestroy {
     link: string;
```
Comment (YuraRiabov) [inline frontend/src/app/modules/external-booking/external-booking-choose-time-page/external-booking-choose-time-page.component.ts]: Why do you use on destroy, BaseComponent already implements it?

File: frontend/src/app/modules/external-booking/external-booking-choose-time-page/external-booking-choose-time-page.component.ts [hunk 2]
```diff
@@ -243,12 +260,23 @@ export class ExternalBookingTimeComponent extends BaseComponent implements OnIni
...
 
     redirectToChooseMeeting() {
         this.reloadData.emit(this.userLink);
         this.router.navigateByUrl(`/external-booking/choose-meeting/${this.userLink}`);
     }
+
+    override ngOnDestroy(): void {
+        super.ngOnDestroy();
+    }
 }
```
Comment (YuraRiabov) [inline frontend/src/app/modules/external-booking/external-booking-choose-time-page/external-booking-choose-time-page.component.ts]: This does nothing as far as I'm concerned

**Review by YuraRiabov — CHANGES_REQUESTED** [review]

File: frontend/src/app/modules/external-booking/external-booking-choose-time-page/external-booking-choose-time-page.component.ts (hunk 2 above)
Comment (va-dym-ko-val) [reply frontend/src/app/modules/external-booking/external-booking-choose-time-page/external-booking-choose-time-page.component.ts]: Fixed

File: frontend/src/app/modules/external-booking/external-booking-choose-time-page/external-booking-choose-time-page.component.ts (hunk 1 above)
Comment (va-dym-ko-val) [reply frontend/src/app/modules/external-booking/external-booking-choose-time-page/external-booking-choose-time-page.component.ts]: Fixed

**Review by YuraRiabov — APPROVED** [review]

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs [hunk 3]
```diff
@@ -325,6 +325,12 @@ public async Task<bool> ValidateLinkAsync(long? slotId, string slotLink)
             return !await _context.AvailabilitySlots.AnyAsync(s => s.Id != slotId && s.Link == slotLink);
         }
 
+        public async Task<bool> ValidateSlotPasswordAsync(string slotLink, string password)
+        {
+            var slot = await _context.AvailabilitySlots.FirstOrDefaultAsync(el => el.Link == slotLink) ?? throw new KeyNotFoundException("Availability slot doesn't exist");
+            return slot?.PasswordProtection == password;
```
Comment (tatianahutii) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs]: You can combine these lines and use AnyAsync

File: frontend/src/app/shared/components/slot-password-window/slot-password-window.component.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        @Inject(MAT_DIALOG_DATA) public data: ISlotPasswordData,
+        private dialogRef: MatDialogRef<SlotPasswordWindowComponent>,
+        private router: Router,
+        private availabilitySlotService: AvailabilitySlotService,
+        private notificationService: NotificationService,
+    ) {
+        this.title = data.title;
+        this.message = data.message;
+        this.buttonsOptions = data.buttonsOptions;
+        this.slotLink = data.slotLink;
```
Comment (tatianahutii) [inline frontend/src/app/shared/components/slot-password-window/slot-password-window.component.ts]: Please create and use one interface

File: frontend/src/app/shared/components/slot-password-window/slot-password-window.component.ts [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+                class: 'confirm-accept-button',
+                label: 'Ok',
+                onClickEvent: this.enterPasswordEventEmitter,
+            },
+            {
+                class: 'confirm-cancel-button',
+                label: 'Cancel',
+                onClickEvent: this.leaveEventEmitter,
+            },
+        ];
```
Comment (tatianahutii) [inline frontend/src/app/shared/components/slot-password-window/slot-password-window.component.ts]: Probably this action do not match with method name

File: frontend/src/app/shared/components/slot-password-window/slot-password-window.component.ts [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        this.addEventsSubscription();
+    }
+
+    onClick(event: EventEmitter<void>) {
+        event?.next();
+    }
+
+    private addEventsSubscription() {
+        this.leaveEventSubscription = this.leaveEventEmitter.subscribe(() => this.leaveSlotPasswordWindow());
+        this.enterPasswordEventSubscription = this.enterPasswordEventEmitter.subscribe(() => this.enterPassword());
```
Comment (tatianahutii) [inline frontend/src/app/shared/components/slot-password-window/slot-password-window.component.ts]: Why do you need these subscriptions?

**Review by tatianahutii — CHANGES_REQUESTED** [review]

**Review by tatianahutii — APPROVED** [review]

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs [hunk 7]
```diff
@@ -325,6 +325,11 @@ public async Task<bool> ValidateLinkAsync(long? slotId, string slotLink)
             return !await _context.AvailabilitySlots.AnyAsync(s => s.Id != slotId && s.Link == slotLink);
         }
 
+        public async Task<bool> ValidateSlotPasswordAsync(string slotLink, string password)
+        {
+            return await _context.AvailabilitySlots.AnyAsync(el => el.Link == slotLink && el.PasswordProtection == password);
+        }
```
Comment (apofis96) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs]: Return task without async-await

**Review by apofis96 — CHANGES_REQUESTED** [review]

**Review by apofis96 — APPROVED** [review]

**Pull request merged**